### PR TITLE
DL-2268 Auth-client upgrade for ATED-SUB-FE

### PR DESCRIPTION
**Pull request by jmountain-hmrc**
# DL-2268 Auth-client upgrade for ATED-SUB-FE

**Auth upgrade**

* Upgraded ated-subscription-frontend to use new auth-client libraries for authentication.

## Checklist

 - [x]  I've made every effort to commit high quality, clean code and I have executed relevant static analyses to be sure
 - [x]  I've included appropriate tests with any code I've added (Unit, Integration, Acceptance etc.)
 - [x]  I've executed the acceptance test pack locally to ensure there are no functional regressions
 - [x]  I've added my code using logical, atomic commits, squashing as appropriate - including the JIRA issue number in the commit message
 - [x]  I've run a dependency check to ensure all dependencies are up to date

### CODE REVIEW

File: app/controllers/auth/AuthFunctionality.scala [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+          Retrievals.allEnrolments and
+          Retrievals.agentCode and
+          Retrievals.credentials
+      ) {
+        case credRole ~ affinityGroup ~ authorisedEnrolments ~ agentCode ~ credentials =>
+          body(AtedSubscriptionAuthData(
+            credRole,
+            affinityGroup,
+            agentCode,
+            credentials.map(creds => UrlSafe.hash(creds.providerId)),
```
Comment (Ufuoma-Daniel) [inline app/controllers/auth/AuthFunctionality.scala]: Doesn't the cred needs to be checked first? It's provider type should be ??? before you collect the providerId.

File: app/utils/AuthUtils.scala [hunk 2]
```diff
@@ -16,32 +16,39 @@
...
-  def isAgent(implicit user: AuthContext): Boolean = user.principal.accounts.agent.isDefined
+  def isAgent(implicit user: AtedSubscriptionAuthData): Boolean =
+    user.enrolments.getEnrolment("HMRC-AGENT-AGENT").isDefined || user.affinityGroup.contains(AffinityGroup.Agent)
 
-  def isAgentAssistant(implicit user: AuthContext): Boolean = user.principal.accounts.agent.map(_.agentUserRole).contains(AgentAssistant)
+  def isAgentAssistant(implicit user: AtedSubscriptionAuthData): Boolean = user.credentialRole.contains(Assistant)
 
-  def isAgentAdmin(implicit user: AuthContext): Boolean = user.principal.accounts.agent.map(_.agentUserRole).contains(AgentAdmin)
+  def isAgentAdmin(implicit user: AtedSubscriptionAuthData): Boolean =
+    user.credentialRole.contains(Admin) || isAgent && user.credentialRole.contains(User)
```
Comment (Ufuoma-Daniel) [inline app/utils/AuthUtils.scala]: @deprecated("Admin and User are equivalent. In the future Admin will be removed. Use User instead!")

File: app/utils/AuthUtils.scala [hunk 3]
```diff
@@ -16,32 +16,39 @@
...
 
-  def agentLink(implicit user: AuthContext): String = {
-    user.principal.accounts.agent.map(_.link).getOrElse {
+  def agentLink(implicit user: AtedSubscriptionAuthData): String = {
+    user.agentCode.map(str => s"/agent/$str").getOrElse {
       Logger.warn(s"[AuthUtils][getAgentLink] Exception - User does not have the correct authorisation ")
       throw new RuntimeException("User is not agent")
```
Comment (Ufuoma-Daniel) [inline app/utils/AuthUtils.scala]: Unhandled exception when no agent code is found.

File: app/utils/AuthUtils.scala [hunk 4]
```diff
@@ -16,32 +16,39 @@
...
 
-  def getAuthLink(implicit user: AuthContext): String = {
-    if (isAgent) agentLink
-    else user.principal.accounts.org.map(_.link).getOrElse(throw new RuntimeException("User is not org"))
+  def getAuthLink(implicit user: AtedSubscriptionAuthData): String = {
+    if (isAgent) {
+      agentLink
+    } else {
+      user.hashedCredId.map(str => s"/org/$str")
+        .getOrElse(throw new RuntimeException("User is not org"))
```
Comment (Ufuoma-Daniel) [inline app/utils/AuthUtils.scala]: Unhandled exception.

File: app/services/RegisteredBusinessService.scala [hunk 5]
```diff
@@ -88,7 +89,8 @@ trait RegisteredBusinessService {
       }
     }
 
-    val agentArn = user.principal.accounts.agent.flatMap(_.agentBusinessUtr.map(_.value))
+    val agentArn = user.enrolments.getEnrolment("HMRC-AGENT-AGENT")
```
Comment (Ufuoma-Daniel) [inline app/services/RegisteredBusinessService.scala]: You made a getArn function that could replace this (and other similar instances of this)

File: app/controllers/auth/AuthFunctionality.scala [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import uk.gov.hmrc.auth.core._
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals
+import uk.gov.hmrc.auth.core.retrieve.~
+import uk.gov.hmrc.http.HeaderCarrier
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait AuthFunctionality extends AuthorisedFunctions {
+
+  val agentEnrolment = "HMRC-AGENT-AGENT"
```
Comment (Ufuoma-Daniel) [inline app/controllers/auth/AuthFunctionality.scala]: Unused.

File: app/controllers/auth/AuthFunctionality.scala (hunk 1 above)
Comment (Ufuoma-Daniel) [reply app/controllers/auth/AuthFunctionality.scala]: No longer relevant now that AuthProvider check has been included.

File: test/controllers/auth/AuthFunctionalitySpec.scala [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+        implicit val fq: FakeRequest[AnyContentAsEmpty.type] = FakeRequest()
+        implicit val messages: Messages = mock[Messages]
+
+        val res: Future[Result] = controllerHarness.authoriseFor(func)
+        status(res) shouldBe 200
+      }
+    }
+  }
+
```
Comment (Ufuoma-Daniel) [inline test/controllers/auth/AuthFunctionalitySpec.scala]: You only have a happy path test case here. The two recover block scenarios are missing.

**Comment by Ufuoma-Daniel** [comment]
Can you updated the pull request template please?

File: app/utils/AuthUtils.scala (hunk 3 above)
Comment (jmountain-hmrc) [reply app/utils/AuthUtils.scala]: Preserving old functionality

File: app/utils/AuthUtils.scala (hunk 4 above)
Comment (jmountain-hmrc) [reply app/utils/AuthUtils.scala]: Preserving old functionality

File: app/controllers/auth/AuthFunctionality.scala (hunk 6 above)
Comment (jmountain-hmrc) [reply app/controllers/auth/AuthFunctionality.scala]: Is it?

Comment (Ufuoma-Daniel) [reply app/controllers/auth/AuthFunctionality.scala]: It is now, it wasn't when I put the comment.

**Pull request merged**